### ISSUE TITLE
Error when trying to add a file to a Builder object

### ISSUE THREAD

**Issue by tandy-1000** [StefanSalewski/gintro#91]
I've tried to do this using `addFromFile` and `newBuilderFromFile` but it gives the same error, I have also tried using the absolute path to the `.ui` file.
This issue exists on both `stable` and `devel` versions of nim.

```
Hint: 277525 LOC; 25.448 sec; 339.754MiB peakmem; Release build; proj: /home/tandy/Programming/Nim/moodfm/src/main.nim; out: /home/tandy/Programming/Nim/moodfm/src/main [SuccessX]
Hint: /home/tandy/Programming/Nim/moodfm/src/main  [Exec]

(process:156891): Gtk-CRITICAL **: 18:05:25.852: _gtk_style_provider_private_get_settings: assertion 'GTK_IS_STYLE_PROVIDER_PRIVATE (provider)' failed

(process:156891): Gtk-CRITICAL **: 18:05:25.852: _gtk_style_provider_private_get_settings: assertion 'GTK_IS_STYLE_PROVIDER_PRIVATE (provider)' failed

(process:156891): Gtk-CRITICAL **: 18:05:25.852: _gtk_style_provider_private_get_settings: assertion 'GTK_IS_STYLE_PROVIDER_PRIVATE (provider)' failed
Traceback (most recent call last)
/home/tandy/Programming/Nim/moodfm/src/main.nim(16) main
/home/tandy/Programming/Nim/moodfm/src/main.nim(8) main
/home/tandy/Programming/Nim/moodfm/src/dialog.nim(20) dialog
/home/tandy/.nimble/pkgs/gintro-0.8.1/gintro/gtk.nim(29593) newBuilderFromFile
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
Error: execution of an external program failed: '/home/tandy/Programming/Nim/moodfm/src/main '
```

**Comment by StefanSalewski** [comment]
Sorry but I have no idea what and how you are trying to do something.

**Comment by tandy-1000** [comment]
[Here](https://gitlab.com/tandy1000/moodfm/-/blob/master/src/dialog.nim) is the proc for the application where the program is loading a Glade file into a Gtk Builder object
```
proc dialog*(app: Application): string =
  let
    builder = newBuilderFromFile("assets/dialog.ui")
    window = builder.getApplicationWindow("dialog")
  var
    entry: Entry = builder.getEntry("userEntry")
    go: Button = builder.getButton("go")
  entry.connect("activate", enterPress, app)
  go.connect("clicked", goClick, (entry, app))
  builder.setApplication(app)
  setCSS(window)
  window.showAll
  return user
```

[Here](https://gitlab.com/tandy1000/moodfm/-/blob/master/src/main.nim) it is being executed.
```
proc main =
  let dialogApp = newApplication("org.gtk.dialogApp")
  var user: string = dialog(dialogApp)
```

Let me know if this helps.

**Comment by StefanSalewski** [comment]
The line

```nim
go.connect("clicked", goClick, (entry, app))
```

looks interesting, as you pass a tuple as the last (optional) argument. I can not remember a case where someone tested passing a tuple. I know var parameters do not work, while basic types like ints work, and I think that we have tested objects and they work. So maybe test for tuples first. Thinking about it, I really expect that tuples will fail, I have not even an idea how the macro code for it may look like. You may see yourself, the connect macros are in file gintro/gintro/gimpl.nim which is copied by nimble install to .nimble/pkgs/gintro-0.8.1/gintro/gimpl.nim.

If you should be not able to fix it yourself, then a full minimal Nim example that we can compile would help. A working C or Python example would be fine too, so that we can convert it to working Nim example.

Unfortunately Dialog examples are still missing from our example collection.

**Comment by tandy-1000** [comment]
I don't think its the tuple thing even thought it might look it, since the code did work for a while like that. I will try get a minimum working example and post it here if the issue persists.

**Comment by StefanSalewski** [comment]
>code did work for a while

So the program does not crash immediately? Then it is generally the GC. The default refc GC remains inactive for a while, so programs may not crash immediately. I generally use GC ARC, as it would crash immediately, so finding bugs is easier.

I think the tuple is a problem, but there may be another bug too. I think tuples should not work currently, I really think I have not cared for them when I wrote the connect macro. And I still have no idea how to handle anonymous tuples, I would have to ask Araq or maybe another Nim core dev. Handling objects was easy, it is only calling GC_ref to avoid that the GC collects the object. But for tuple? There is no plain GC_ref for tuples, so we would have to care for the concrete implementation details.

We may intent to rewrite the whole connect macro, recently someone asked for lambda support, and making var parameters work would be nice too. But I think only a real macro expert can do that.

**Comment by tandy-1000** [comment]
It does crash immediately, but I mean that it used to work with the tuple solution I believe.
I have tried running some previous iterations of the code which if I remember correctly, did run properly, but they don't seem to anymore. Which makes me think the cause may be something like nim / gintro / gtk updating and changing but we will see once I get the min working example ready.

**Comment by StefanSalewski** [comment]
>used to work with the tuple solution

Yes, a tuple (int, int) may work. I have not looked into the gimpl.nim file, as I am busy currently. But I assume when tuples compile at all, then they may be copied. Copy of (int, int) is fine I guess, but copy of (entry, app) is fully wrong and can not work. Maybe interesting to investigate how we can make it work, but I think we have to ask Araq, and Araq does not like gtk and gintro that much. If you want to get it working, why not put entry and app in a ref object? I assume that should work, maybe we have examples for that already. Or ask Araq about the tuple -- if he sees that there are gintro users he may care more.
